src/services/exit_order.py

```diff
@@ -0,0 +1,360 @@
+    StakingModuleId,
+    NodeOperatorId,
+    ValidatorsByNodeOperator,
+)
+from src.web3py.typings import Web3
+
+
+@dataclass
+class NodeOperatorPredictableState:
+    predictable_validators_total_age: int
+    predictable_validators_count: int
+    targeted_validators_limit_count: int | None
+    delayed_validators_count: int
+
+
+@dataclass
+class ValidatorToExitIteratorConfig:
+    max_validators_to_exit: int
+    validator_delayed_timeout_in_slots: int
+
+
+class ValidatorToExitIterator:
+    """
+    Exit order predicates sequence:
+    1. Validator whose operator with the lowest number of delayed validators
+    2. Validator whose operator with the highest number of targeted validators to exit
+    3. Validator whose operator with the highest stake weight
+    4. Validator whose operator with the highest number of validators
+    5. Validator with the lowest index
+    """
+    v_conf: ValidatorToExitIteratorConfig
+    staking_module_id: dict[Address, StakingModuleId]
+    exitable_lido_validators: list[LidoValidator]
+    lido_node_operator_stats: dict[NodeOperatorGlobalIndex, NodeOperatorPredictableState]
+    total_predictable_validators_count: int
+    left_queue_count: int = 0
+
+    def __init__(
+        self,
+        w3: Web3,
+        blockstamp: ReferenceBlockStamp,
+        c_conf: ChainConfig,
+    ):
+        self.w3 = w3
+        self.blockstamp = blockstamp
+        self.c_conf = c_conf
+
+    def __iter__(self) -> Iterable[tuple[NodeOperatorGlobalIndex, LidoValidator]]:
+        """
+        Prepare queue state for the iteration:.
+        Determine exitable Lido validators and collect operators stats to sort exitable validators
+        """
+        self.left_queue_count = 0
+
+        self._get_config()
+
+        lido_validators = {
+            v.validator.pubkey: v for v in self.w3.lido_validators.get_lido_validators(self.blockstamp)
+        }
+
+        operators = self.w3.lido_validators.get_lido_node_operators(self.blockstamp)
+        operator_validators = self.w3.lido_validators.get_lido_validators_by_node_operators(self.blockstamp)
+        self.staking_module_id = {
+            operator.staking_module.staking_module_address: operator.staking_module.id
+            for operator in operators
+        }
+
+        # DRY: code duplication (LidoValidatorStateService)
+        last_requested_to_exit_indices_per_operator = self._get_last_requested_to_exit_indices(
+            operator_validators.keys()
+        )
+
+        self.exitable_lido_validators = list(
+            self._get_exitable_lido_validators(operator_validators, last_requested_to_exit_indices_per_operator)
+        )
+
+        self.lido_node_operator_stats = self._prepare_lido_node_operator_stats(
+            operators, operator_validators, last_requested_to_exit_indices_per_operator
+        )
+
+        not_lido_predictable_validators_count = len([
+            v for v in self.w3.cc.get_validators(self.blockstamp)
```

Comment:
maybe choose a single style on list comprehensions vs `sum(map(...` vs `reduce` (though, don't want `reduce` at all)

Reply:
will be fixed during refactoring